File: 2016/3.cpp
#include <iostream>
#include <queue>
#include <cmath>
using namespace std;
#define ll long long

bool checkPrime(ll n) {
    if (n == 2)
        return 1;
    if (!(n & 1))
        return 0;
    for (int i = 3; i <= ceil(sqrt(n)); i+=2)
        if (!(n % i))
            return 0;
    return 1;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    ll l, p, q;
    cin >> l >> p >> q;

    vector<ll> shortest(l+1);
    for (int i = 0; i <= l; i++)
        shortest[i] = 1e9;
    shortest[p] = 1;
    vector<ll> powers;
    for (int i = 1; i <= l; i <<= 1)
        powers.push_back(i);
    queue<pair<ll,ll>> processing;
    processing.push({p,1});
    while (!processing.empty()) {
        pair<ll,ll> current = processing.front();
        processing.pop();
        if (current.first == q) {
            cout << current.second << "\n";
            return 0;
        }
        for (int i = 0; i < powers.size(); i++) {
            if (current.first - powers[i] > 0) {
                if (current.second + 1 < shortest[current.first-powers[i]]) {
                    if (checkPrime(current.first - powers[i])) {
                        processing.push({current.first-powers[i], current.second+1});
                        shortest[current.first-powers[i]] = current.second+1;
                    }
                }
            } if (current.first + powers[i] <= l) {
                if (current.second + 1 < shortest[current.first+powers[i]]) {
                    if (checkPrime(current.first + powers[i])) {
                        processing.push({current.first + powers[i], current.second + 1});
                        shortest[current.first+powers[i]] = current.second + 1;
                    }
                }
            }
        }
    }
}

File: 2016/3.py
from math import ceil
from collections import deque
l, p, q = [int(i) for i in input().split()]
# Start is p, end is q
# I think that this is just a bfs
# Currently works but is just very slow, maybe I should implement dijkstra's logic
# This program works but, it takes just a little too long with larger inputs
# So, I rewrote the whole thing in cpp in "3.cpp" and it runs even the bigger inputs in about a second

# This checked dictionary caches the output for numbers that were already run through checkPrime
checked = dict()
def checkPrime(number):
    if number == 2:
        return 1
    if not (number & 1):
        return 0
    if checked.get(number) != None:
        return checked[number]
    for i in range(3, ceil(pow(number,0.5)) + 1, 2):
        if not (number % i):
            checked[number] = 0
            return 0
    checked[number] = 1
    return 1

shortest = [int(1e9) for i in range(l+1)]
powers = [pow(2,i) for i in range(24)]
queue = deque([[p,1]])
while len(queue) > 0:
    current = queue.popleft()
    if current[0] == q:
        print(current[1])
        break
    for i in powers:
        if current[0] - i > 0:
            if current[1] + 1 < shortest[current[0]-i]:
                if checkPrime(current[0]-i):
                    queue.append([current[0]-i,current[1]+1])
                    shortest[current[0]-i] = current[1]+1
        if current[0] + i <= l:
            if current[1] + 1 < shortest[current[0]+i]:
                if checkPrime(current[0]+i):
                    queue.append([current[0]+i,current[1]+1])
                    shortest[current[0]+i] = current[1] + 1
